File: StacksQueues&Deques/Quene-UsingTwoStacks/JS/Quene-UsingTwoStacks.js
/*
Implement a Queue - Using Two Stacks
Given the Stack class below, implement a Queue class using two stacks! 
Note, this is a "classic" interview problem. Use a Python list data structure as your Stack.
*/

var Queue2Stacks = function() {

  this.inStack=[];
  this.outStack=[];

}

Queue2Stacks.prototype.enqueue = function (value){
  this.inStack.push(value);
}

Queue2Stacks.prototype.dequeue = function (){
  const len = this.inStack.length;

  if (this.outStack.length === 0) {
    for(var i = 0; i < len; i++){
      this.outStack.push(this.inStack.pop());
    }
  } 
  
  return this.outStack.pop();
}


q = new Queue2Stacks();

for (var j =0; j< 5;j++){
  q.enqueue(j);
}  

for (var k =0; k< 5;k++){
  console.log(q.dequeue());
}  

File: StacksQueues&Deques/Quene-UsingTwoStacks/python/Quene-UsingTwoStacks.py
# -*- coding: utf-8 -*-
"""
Implement a Queue - Using Two Stacks
Given the Stack class below, implement a Queue class using two stacks! 
Note, this is a "classic" interview problem. Use a Python list data structure as your Stack.
"""

class Queue2Stacks(object):
    
  def __init__(self):
    # Two Stacks
    self.in_stack = []
    self.out_stack = []
    
  def enqueue(self, element):
    # FILL OUT CODE HERE
    self.in_stack.append(element)
    pass
  
  def dequeue(self):
    # FILL OUT CODE HERE
    if not self.out_stack:
      while self.in_stack:
          self.out_stack.append(self.in_stack.pop())
    return self.out_stack.pop()
    pass

q = Queue2Stacks()

for i in range(5):
  q.enqueue(i)
    
for i in range(5):
  print (q.dequeue())